public/jira/rollup/rollup.ts

```diff
@@ -0,0 +1,448 @@
+/**
+ * This module provides utilities for building and processing reporting hierarchies of derived issues and releases.
+ * It includes functions to group issues by hierarchy levels or types, build parent-child relationships,
+ * and perform rollup calculations. The rollup functions enable aggregating data from child issues up through
+ * the hierarchy, allowing for customized data summaries and metadata generation at each hierarchy level.
+ */
+import type { DerivedIssue } from "../derived/derive";
+
+import type { DerivedRelease } from "../releases/derive";
+
+export type IssueOrRelease<CustomFields = unknown> =
+  | (DerivedIssue & CustomFields)
+  | (DerivedRelease & CustomFields);
+
+interface ReportingHierarchy {
+  depth: number;
+  childKeys: string[];
+  parentKeys: string[];
+}
+
+type ReportingHierarchyIssueOrRelease<CustomFields> =
+  IssueOrRelease<CustomFields> & {
+    reportingHierarchy: ReportingHierarchy;
+  };
+
+export interface RollupResponseItem<Data, Meta> {
+  metadata: Meta;
```

Comment:
I removed the `RollupData<Data>` and `RollupMetadata<Meta>` fields
```
export type RollupData<D> = D & Record<string, any>;
export type RollupMetadata<M> = M & Record<string, any>;
```

And updated this
```
export interface RollupResponseItem<D, M> {
  metadata: RollupMetadata<M>;
  rollupData: RollupData<D>[];
}
```
to directly use `Data` and `Meta`
```
export interface RollupResponseItem<Data, Meta> {
  metadata: Meta;
  rollupData: Data[];
}
```
This seems to work fine after I made sure that the 3 generics are being passed down as needed from the main functions to all of the internal ones

Reply:
There should not be any instance of `any` in the code

Reply:
Is Data always going to be an array?

Reply:
Yes should be. 
All of the existing tests data expects rollupData to be an array. Including the tests of the custom rollup functions like `warning-issues`, `blocked-status-issues`, etc..

If a new rollup function was added that expects a different format, we can update it then. But for now type array aligns with the existing examples